### PR TITLE
Feature/service template

### LINKED ISSUE #45: Add entity service template
Preferred usage:
dotnet new logofx-entity-service -n `{entity-name}` 
Desired outcome:
- [x] entity service contract named `I{entity-name}Service` is added to the contracts project.  The contract should contain CRUD asynchronous methods' definitions in terms of model entity.
model service explicit implementation named `{entity-name}Service` is added to the implementations project.
- [x] The implementation should contain CRUD asynchronous methods' implementations in terms of model entity with basic functionality.
- [x] This would mean also invoking the template for adding a new model entity via CLI (depends on #44 completion)
The model service implementation should include data provider injection. This would mean that the outcome would include the following code files to be added:
- [x] data provider contract named `I{entity-name}DataProvider` is added to the data provider contracts project. The contract should contain CRUD synchronous methods' definitions in terms of DTO.
- [x] fake data container named `{entity-name}DataContainer` is added to the fake data containers project. See the example of code inside the WPF template.
- [x] fake data provider builder named `{entity-name}ProviderBuilder` is added to the fake data provider builders project. The implementation should contain setup for all CRUD synchronous methods defined in the data contract provider
- [x] fake data provider named `Fake{entity-name}Provider` is added to the fake data providers project. See the example of code inside the WPF template.
- [x] real data provider named `Real{entity-name}Provider` is added to the real data providers project. The implementation should be compile-only, i.e. without any actual implementation.

There are also existing registration files and since we use dynamic loading and .NET/.NET Core we have to register the dependencies explicitly. This means updating the correspondent `Module.cs` files where needed:
- [x] The following code (or equivalent) is added to the `Fake\Module.cs`file 
```csharp

public void RegisterModule(IDependencyRegistrator dependencyRegistrator)
        {
            dependencyRegistrator
...
                .AddInstance(InitializeSampleContainer())
                .AddSingleton<ISampleProvider, FakeSampleProvider>()
...
          ;
...
            dependencyRegistrator.RegisterInstance(SampleProviderBuilder.CreateBuilder());
...
        }
...
        private static ISampleContainer InitializeSampleContainer()
        {
            var sampleContainer = new SampleContainer();
            sampleContainer.UpdateItems(new[]
            {
                new SampleItemDto
                {
                    Id = Guid.NewGuid(),
                    DisplayName = "PC",
                    Value = 8
                },

                new SampleItemDto
                {
                    Id = Guid.NewGuid(),
                    DisplayName = "Acme",
                    Value = 10
                },

                new SampleItemDto
                {
                    Id = Guid.NewGuid(),
                    DisplayName = "Bacme",
                    Value = 3
                },

                new SampleItemDto
                {
                    Id = Guid.NewGuid(),
                    DisplayName = "Exceed",
                    Value = 100
                },

                new SampleItemDto
                {
                    Id = Guid.NewGuid(),
                    DisplayName = "Acme2",
                    Value = 10
                }
            });
            return sampleContainer;
        }
...
```

Tasks:

- [x] #65
- [x] #72 
- [x] #73 
- [x] #74 
- [x] #75 
- [x] #76 
- [x] #77 
- [x] #78 
- [x] #79 
- [x] #80

### PR DESCRIPTION
**Pull request by dmitrybublik**
Related to #45

### CODE REVIEW

File: common/Common.Infra/ProcessExtensions.cs [hunk 1]
```diff
@@ -3,11 +3,30 @@
 using System.Diagnostics;
 using System.IO;
 using System.Management;
+using System.Threading;
```
Comment (godrose) [inline common/Common.Infra/ProcessExtensions.cs:6]: Do you need some kind of Process Manager here?

File: devops/copy-template-logofx-entity.cmd [hunk 2]
```diff
@@ -6,7 +6,7 @@ if %ERRORLEVEL% NEQ 0 (
 	goto EXIT
 )
 
-call generate-utils.cmd LogoFX.Templates.Model.Entity
+call generate-utils.cmd LogoFX.Templates.Model.Entity model-utils
```
Comment (godrose) [inline devops/copy-template-logofx-entity.cmd]: Shouldn't we use `entity-utils`? See `service-utils` naming convention

File: specs/LogoFX.Cli.Dotnet.Specs.Features/Service Generation.feature [hunk 3]
```diff
@@ -12,9 +12,28 @@ Scenario Outline: Generate service entity in an empty folder
 	| -n                   | <entityNameValue> |
 	| <solutionNameOption> | Test              |
 	| --allow-scripts      | yes               |
-	Then The folder 'Generation' contains generated service entity objects for name '<entityNameValue>' for solution name 'Test'
+	Then The folder 'Generation' contains generated model entity objects for name '<entityNameValue>' for solution name 'Test'
+	And The folder 'Generation' contains generated service entity objects for name '<entityNameValue>' for solution name 'Test'
```
Comment (godrose) [inline specs/LogoFX.Cli.Dotnet.Specs.Features/Service Generation.feature]: It should be 'entity service'

File: specs/LogoFX.Cli.Dotnet.Specs.Features/Service Generation.feature [hunk 4]
```diff
@@ -12,9 +12,28 @@ Scenario Outline: Generate service entity in an empty folder
 	| -n                   | <entityNameValue> |
 	| <solutionNameOption> | Test              |
 	| --allow-scripts      | yes               |
-	Then The folder 'Generation' contains generated service entity objects for name '<entityNameValue>' for solution name 'Test'
+	Then The folder 'Generation' contains generated model entity objects for name '<entityNameValue>' for solution name 'Test'
```
Comment (godrose) [inline specs/LogoFX.Cli.Dotnet.Specs.Features/Service Generation.feature]: It should be 'entity'

File: specs/LogoFX.Cli.Dotnet.Specs.Features/Service Generation.feature [hunk 5]
```diff
@@ -12,9 +12,28 @@ Scenario Outline: Generate service entity in an empty folder
...
+	And The folder 'Generation' contains generated service entity objects for name '<entityNameValue>' for solution name 'Test'
 
 	Examples:
 	| entityNameValue | solutionNameOption |
 	| Sample          | -sn                |
 	| Another         | --solution-name    |
+
+Scenario Outline: Generate service entity in existing solution
```
Comment (godrose) [inline specs/LogoFX.Cli.Dotnet.Specs.Features/Service Generation.feature]: It should be entity service

File: specs/LogoFX.Cli.Dotnet.Specs.Features/Service Generation.feature [hunk 6]
```diff
@@ -12,9 +12,28 @@ Scenario Outline: Generate service entity in an empty folder
...
+	When I install the template pack 'LogoFX.Templates' from local package
+	And I create a folder named 'Generation'
+	And I generate the code in folder named 'Generation' using 'logofx-wpf' template with the default options
+	And I generate the code in folder named 'Generation' using 'logofx-entity-service' template with the following options
+	| Name                 | Value  |
+	| -n                   | Sample |
+	| <solutionNameOption> | Test   |
+	| --allow-scripts      | yes    |
+	Then The folder 'Generation' contains working LogoFX template-based solution
+	And The folder 'Generation' contains generated model entity objects for name 'Sample' for solution name 'Test'
```
Comment (godrose) [inline specs/LogoFX.Cli.Dotnet.Specs.Features/Service Generation.feature]: It should be entity

File: specs/LogoFX.Cli.Dotnet.Specs.Features/Service Generation.feature [hunk 7]
```diff
@@ -12,9 +12,28 @@ Scenario Outline: Generate service entity in an empty folder
...
+	And The folder 'Generation' contains generated service entity objects for name 'Sample' for solution name 'Test'
```
Comment (godrose) [inline specs/LogoFX.Cli.Dotnet.Specs.Features/Service Generation.feature]: It should be entity service

File: specs/LogoFX.Cli.Dotnet.Specs.Steps/AssertionHelper.cs [hunk 8]
```diff
@@ -9,7 +11,16 @@ internal static class AssertionHelper
 
         internal static void AssertGeneratedCode(this GeneratedFolder structure)
         {
-            Directory.Exists(Path.Combine(structure.RootPath, structure.Name)).Should().BeTrue();
+            var path = Path.Combine(structure.RootPath, structure.Name);
+            var count = 10;
```
Comment (godrose) [inline specs/LogoFX.Cli.Dotnet.Specs.Steps/AssertionHelper.cs:16]: Add TODO: Get from config.json

File: specs/LogoFX.Cli.Dotnet.Specs.Steps/AssertionHelper.cs [hunk 9]
```diff
@@ -9,7 +11,16 @@ internal static class AssertionHelper
...
+            while (!Directory.Exists(path) && count > 0)
+            {
+                Debug.WriteLine($"Checking '{path}, attempts: {count}'");
+                count -= 1;
+                Thread.Sleep(2000);
```
Comment (godrose) [inline specs/LogoFX.Cli.Dotnet.Specs.Steps/AssertionHelper.cs:22]: Add TODO: Get from config.json

Comment (godrose) [inline specs/LogoFX.Cli.Dotnet.Specs.Steps/AssertionHelper.cs:22]: What is the recommended way: `Thread.Sleep` or `Task.Delay`? 🤔

File: specs/LogoFX.Cli.Dotnet.Specs.Steps/GenerationFolderExtensions.cs [hunk 10]
```diff
@@ -141,13 +141,13 @@ public static class AssemblyInfo
     {{
         public static string AssemblyName {{ get; }} = $""{{Assembly.GetExecutingAssembly().GetName().Name}}.dll"";
     }}
-}}").WithFile("ISampleProvider.cs", $@"using System;
+}}").WithFile("ISampleDataProvider.cs", $@"using System;
```
Comment (godrose) [inline specs/LogoFX.Cli.Dotnet.Specs.Steps/GenerationFolderExtensions.cs:146]: What was the intention of adding `Data` part? If I recall correctly the current convention is `I{dto-name}Provider`

File: specs/LogoFX.Cli.Dotnet.Specs.Steps/CodeGenerationSteps.cs [hunk 11]
```diff
@@ -33,7 +32,7 @@ private void ExecuteTemplate(string folderName, string shortName, IEnumerable<Ar
 
             var optionsLine = options == null ? null : string.Join(" ", options.Select(k => $"{k.Name}  {k.Value}"));
             var args = $"new {shortName} {optionsLine}";
-            var execInfo = _processManagementService.Start(Path.Combine(path, "dotnet"), args, Consts.ProcessExecutionTimeout);
+            var execInfo = _processManagementService.Start(Path.Combine(path, "dotnet"), args);
```
Comment (godrose) [inline specs/LogoFX.Cli.Dotnet.Specs.Steps/CodeGenerationSteps.cs:36]: Add TODO: Needs dedicated documentation

File: specs/LogoFX.Cli.Dotnet.Specs.Tests.Contracts/IProcessManagementService.cs [hunk 12]
```diff
@@ -2,7 +2,7 @@
 {
     public interface IProcessManagementService
     {
-        ExecutionInfo Start(string tool, string args, int? waitTime = null);
+        ExecutionInfo Start(string tool, string args, int? pause = 2000);
```
Comment (godrose) [inline specs/LogoFX.Cli.Dotnet.Specs.Tests.Contracts/IProcessManagementService.cs:6]: Add TODO: get from config.json. Use same source across all usages

File: specs/LogoFX.Cli.Dotnet.Specs.Tests.Infra/WindowsProcessManagementService.cs [hunk 13]
```diff
@@ -26,68 +25,21 @@ public ExecutionInfo Start(string tool, string args, int? waitTime = null)
...
-                else
-                {
-                    process.WaitForExit();
-                }
 
-                if (!process.HasExited)
+                if (pause.HasValue && !exitInfo.IsError)
                 {
-                    process.KillProcessAndChildren();
+                    Thread.Sleep(pause.Value);
```
Comment (godrose) [inline specs/LogoFX.Cli.Dotnet.Specs.Tests.Infra/WindowsProcessManagementService.cs:41]: See previous comments regarding `Thread.Sleep`

File: specs/LogoFX.Cli.Dotnet.Specs.Tests.Infra/WindowsProcessManagementService.cs [hunk 14]
```diff
@@ -11,7 +10,7 @@ namespace LogoFX.Cli.Dotnet.Specs.Tests.Infra
     [UsedImplicitly]
     internal sealed class WindowsProcessManagementService : IProcessManagementService
     {
-        public ExecutionInfo Start(string tool, string args, int? waitTime = null)
+        public ExecutionInfo Start(string tool, string args, int? pause = 2000)
```
Comment (godrose) [inline specs/LogoFX.Cli.Dotnet.Specs.Tests.Infra/WindowsProcessManagementService.cs:14]: See previous comments regarding time value

File: templates/LogoFX.Templates.Model.Entity/.template.config/template.json [hunk 15]
```diff
@@ -1,16 +1,16 @@
...
-  "groupIdentity":"LogoFX.Model",
-  "shortName": "logofx-model",
+  "groupIdentity":"LogoFX.Entity",
+  "shortName": "logofx-entity",
   "sourceName": "SampleModel",
   "sources": [
     {
       "copyOnly": [
-        "**/utils/**"
+        "**/model-utils/**"
```
Comment (godrose) [inline templates/LogoFX.Templates.Model.Entity/.template.config/template.json]: Shouldn't it be `entity-utils`?

File: templates/LogoFX.Templates.Model.Entity/setup.cmd [hunk 16]
```diff
@@ -1,14 +1,18 @@
 REM Modify the code
 
-dotnet .\utils\ModifyTool.dll .\LogoFX.Templates --entity SampleModel
+dotnet .\model-utils\ModifyTool.dll .\LogoFX.Templates --model SampleModel
```
Comment (godrose) [inline templates/LogoFX.Templates.Model.Entity/setup.cmd]: Shouldn't it be `entity-utils` and `--entity`?

File: templates/LogoFX.Templates.Service.Entity/LogoFX.Templates.Data.Fake.ProviderBuilders/SampleModelProviderBuilder.cs [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+namespace LogoFX.Templates.Data.Fake.ProviderBuilders
+{
+    public sealed class SampleModelProviderBuilder : FakeBuilderBase<ISampleModelDataProvider>.WithInitialSetup
+    {
+        private readonly List<SampleModelDto> _itemsStorage = new List<SampleModelDto>();
+
+        private SampleModelProviderBuilder()
+        {
+
```
Comment (godrose) [inline templates/LogoFX.Templates.Service.Entity/LogoFX.Templates.Data.Fake.ProviderBuilders/SampleModelProviderBuilder.cs]: Please fix this `codefactor` check notice

File: templates/LogoFX.Templates.WPF/LogoFX.Templates.WPF.Data.Fake.ProviderBuilders/SampleProviderBuilder.cs [hunk 18]
```diff
@@ -9,7 +9,7 @@
 
 namespace LogoFX.Templates.WPF.Data.Fake.ProviderBuilders
 {
-    public sealed class SampleProviderBuilder : FakeBuilderBase<ISampleProvider>.WithInitialSetup
+    public sealed class SampleProviderBuilder : FakeBuilderBase<ISampleDataProvider>.WithInitialSetup
```
Comment (godrose) [inline templates/LogoFX.Templates.WPF/LogoFX.Templates.WPF.Data.Fake.ProviderBuilders/SampleProviderBuilder.cs:12]: Again, what should be the  convention? `I{dto-name}Provider` or `I{dto-name}DataProvider` 🤔

File: templates/LogoFX.Templates.Service.Entity/LogoFX.Templates.Data.Real.Providers/SampleModelDataProvider.cs [hunk 19]
```diff
@@ -0,0 +1,30 @@
+﻿using System;
+using System.Collections.Generic;
+using LogoFX.Templates.Data.Contracts.Dto;
+using LogoFX.Templates.Data.Contracts.Providers;
+
+namespace LogoFX.Templates.Data.Real.Providers
+{
+    internal sealed class SampleModelDataProvider : ISampleModelDataProvider
```
Comment (godrose) [inline templates/LogoFX.Templates.Service.Entity/LogoFX.Templates.Data.Real.Providers/SampleModelDataProvider.cs:9]: Add TODO: Use explicit implementation

File: utils/DevopUtils/UninstallTemplate/Program.cs [hunk 20]
```diff
@@ -209,13 +209,13 @@ private static string FindUninstallString(string[] lines, string name, TemplateN
             return null;
         }
 
-        private static string[] LaunchApp(string launchString)
+        private static IProcessExitInfo LaunchApp(string launchString)
         {
             var strings = launchString.Split(' ');
```
Comment (godrose) [inline utils/DevopUtils/UninstallTemplate/Program.cs:215]: Add TODO: Make this code clearer

File: utils/ModifyTool/Extra/Provider/Module.cs [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+using Solid.Practices.Modularity;
+
+namespace LogoFX.Templates.WPF.Data.Fake.Providers
+{
+    [UsedImplicitly]
+    internal sealed class Module : ICompositionModule<IDependencyRegistrator>
+    {
+        public void RegisterModule(IDependencyRegistrator dependencyRegistrator)
+        {
+
```
Comment (godrose) [inline utils/ModifyTool/Extra/Provider/Module.cs]: Please fix this check notice from `CodeFactor`

File: utils/ModifyTool/Extra/Provider/Module.cs [hunk 22]
```diff
@@ -0,0 +1,20 @@
...
+        }
+    }
```
Comment (godrose) [inline utils/ModifyTool/Extra/Provider/Module.cs:18]: Please fix this check notice from `CodeFactor`

File: utils/ModifyTool/ModelEngine.Module.cs [hunk 23]
```diff
@@ -8,114 +7,35 @@
...
+        /// <param name="entityName"></param>
         [SuppressMessage("ReSharper", "PossibleNullReferenceException")]
         public void RegisterMappers(string entityName)
         {
             var moduleFilePath = Path.Combine(GetProjectFolder(), ModuleFileName);
 
             if (!File.Exists(moduleFilePath))
             {
-                var helper = new ResourceHelper(ModuleFileName);
+                var helper = new ResourceHelper("Model", ModuleFileName);
```
Comment (godrose) [inline utils/ModifyTool/ModelEngine.Module.cs:31]: Add TODO: Refactor helper creation

File: utils/ModifyTool/ModelEngine.cs [hunk 24]
```diff
@@ -0,0 +1,13 @@
+﻿namespace ModifyTool
+{
+    internal sealed partial class ModelEngine : EngineBase
+    {
+        public ModelEngine(string solutionFolder)
+            : base(solutionFolder)
+        {
+
```
Comment (godrose) [inline utils/ModifyTool/ModelEngine.cs]: Please fix this check notice from `CodeFactor`

File: utils/ModifyTool/ModelEngine.cs [hunk 25]
```diff
@@ -0,0 +1,13 @@
...
+        }
+
+        protected override string GetProjectSuffix() => "Model";
+    }
+}
```
Comment (godrose) [inline utils/ModifyTool/ModelEngine.cs:12]: Please fix this check notice from `CodeFactor`

File: utils/ModifyTool/Program.cs [hunk 26]
```diff
@@ -1,41 +1,114 @@
...
+using System.Collections.Generic;
+using Common.Infra;
 
 namespace ModifyTool
 {
-    class Program
+    internal static class Program
     {
-        private const string EntityKey = "entity";
+        private const string ModelKey = "model";
```
Comment (godrose) [inline utils/ModifyTool/Program.cs]: Shouldn't it be 'entity'?

**Review by godrose — CHANGES_REQUESTED** [review]

File: common/Common.Infra/ProcessExtensions.cs (hunk 1 above)
Comment (dmitrybublik) [reply common/Common.Infra/ProcessExtensions.cs:6]: No

File: specs/LogoFX.Cli.Dotnet.Specs.Steps/AssertionHelper.cs (hunk 9 above)
Comment (dmitrybublik) [reply specs/LogoFX.Cli.Dotnet.Specs.Steps/AssertionHelper.cs:22]: Thread.Sleep() pauses current thread. Task.Delay() can do the same, but it used for asynchronous programming with Tasks and must be harder than Thread.Sleep(). Anyway, using Task.Delay() requires more code: Task.Delay(2000).Wait();

File: specs/LogoFX.Cli.Dotnet.Specs.Steps/GenerationFolderExtensions.cs (hunk 10 above)
Comment (dmitrybublik) [reply specs/LogoFX.Cli.Dotnet.Specs.Steps/GenerationFolderExtensions.cs:146]: Wrote //TODO:

File: templates/LogoFX.Templates.WPF/LogoFX.Templates.WPF.Data.Fake.ProviderBuilders/SampleProviderBuilder.cs (hunk 18 above)
Comment (dmitrybublik) [reply templates/LogoFX.Templates.WPF/LogoFX.Templates.WPF.Data.Fake.ProviderBuilders/SampleProviderBuilder.cs:12]: Made //TODO comment

**Review by godrose — APPROVED** [review]

**Pull request merged**